src/ucp/stream/stream_recv.c

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2017.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include <ucp/core/ucp_ep.h>
+#include <ucp/core/ucp_worker.h>
+#include <ucp/core/ucp_context.h>
+#include <ucp/core/ucp_request.h>
+
+#include <ucs/datastruct/mpool.inl>
+#include <ucs/debug/profile.h>
+
+#include <ucp/tag/eager.h> /* TODO: move ucp_eager_sync_hdr_t to common file */
+
+
+typedef struct {
+    uint64_t    uuid;
+} UCS_S_PACKED ucp_stream_am_hdr_t;
+
+UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
+                 (ep, length), ucp_ep_h ep, size_t *length)
+{
+    ucp_recv_desc_t  *rdesc;
+    ucs_status_ptr_t ret;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    if (ucs_queue_is_empty(&ep->stream_data)) {
+        ret = UCS_STATUS_PTR(UCS_OK);
+        goto out;
+    }
+
+    rdesc = ucs_queue_pull_elem_non_empty(&ep->stream_data, ucp_recv_desc_t,
+                                          stream_queue);
+    *length = rdesc->length;
+    ret = (ucs_status_ptr_t)((uintptr_t)(rdesc + 1) + rdesc->hdr_len);
+
+out:
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+
+    return ret;
+}
+
+UCS_PROFILE_FUNC_VOID(ucp_stream_data_release, (ep, data),
+                      ucp_ep_h ep, void *data)
+{
+    ucp_recv_desc_t *rdesc;
+
+    /* TODO: make const header offset */
+    rdesc = (ucp_recv_desc_t *)((uintptr_t)(data -
+                                            sizeof(ucp_stream_am_hdr_t))) - 1;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
+        uct_iface_release_desc((char*)rdesc - sizeof(ucp_eager_sync_hdr_t));
+    } else {
+        ucs_mpool_put_inline(rdesc);
+    }
+
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+}
+
+static ucs_status_t ucp_stream_am_handler(void *arg, void *data, size_t length,
+                                          unsigned am_flags)
+{
+    const size_t            hdr_len = sizeof(ucp_stream_am_hdr_t);
+    ucp_worker_h            worker  = arg;
+    ucp_ep_h                ep;
+    ucp_stream_am_hdr_t     *hdr;
+    ucp_recv_desc_t         *rdesc;
+    size_t                  recv_len;
+    khiter_t                hash_it;
+    ucs_status_t            status;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&worker->mt_lock);
+
+    ucs_assert(length >= hdr_len);
+    recv_len = length - hdr_len;
+
+    /* TODO: merge common code with TAG unexpected */
```

Comment:
i don't think it's possible to merge code because TAG uses hash and multiple queues/lists..

Reply:
This portion of code does not.